File: src/main/java/szewek/flux/tile/Machine2For1Tile.java
package szewek.flux.tile;

import net.minecraft.item.crafting.IRecipeType;
import net.minecraft.tileentity.TileEntityType;
import net.minecraft.util.Direction;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.TranslationTextComponent;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.util.LazyOptional;
import net.minecraftforge.items.CapabilityItemHandler;
import net.minecraftforge.items.IItemHandler;
import net.minecraftforge.items.wrapper.SidedInvWrapper;
import szewek.fl.type.FluxContainerType;
import szewek.flux.container.Machine2For1Container;
import szewek.flux.recipe.AbstractMachineRecipe;

import javax.annotation.Nullable;
import java.util.function.Function;

public final class Machine2For1Tile extends AbstractMachineTile {
	private static final int[] SLOTS_TOP_BOTTOM = {0, 2};
	private static final int[] SLOTS_SIDE = {1, 2};
	private final String titleId;
	private final LazyOptional<? extends IItemHandler>[] sideHandlers = SidedInvWrapper.create(this, Direction.UP, Direction.DOWN, Direction.NORTH);

	public Machine2For1Tile(TileEntityType<?> typeIn, IRecipeType<? extends AbstractMachineRecipe> recipeTypeIn, MenuFactory factory, String titleId) {
		super(typeIn, recipeTypeIn, factory, 2, 1);
		this.titleId = titleId;
	}

	@Override
	public int[] getSlotsForFace(Direction side) {
		switch (side) {
			case UP:
			case DOWN:
				return SLOTS_TOP_BOTTOM;
			default:
				return SLOTS_SIDE;
		}
	}

	@Override
	public ITextComponent getDefaultName() {
		return new TranslationTextComponent(this.titleId);
	}

	@Override
	public <T> LazyOptional<T> getCapability(Capability<T> cap, @Nullable Direction side) {
		if (!removed && cap == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
			if (side == Direction.UP) {
				return sideHandlers[0].cast();
			} else if (side == Direction.DOWN) {
				return sideHandlers[1].cast();
			} else {
				return sideHandlers[2].cast();
			}
		}
		return super.getCapability(cap, side);
	}

	public static Function<TileEntityType<Machine2For1Tile>, Machine2For1Tile> make(final IRecipeType<? extends AbstractMachineRecipe> recipeType, final FluxContainerType<Machine2For1Container> ctype, String titleName) {
		final String titleId = "container.flux." + titleName;
		return type -> new Machine2For1Tile(type, recipeType, (id, player, inv, data) -> new Machine2For1Container(ctype, recipeType, id, player, inv, data), titleId);
	}
}

File: src/main/java/szewek/flux/tile/Machine2For1Tile.kt
package szewek.flux.tile

import net.minecraft.item.crafting.IRecipeType
import net.minecraft.tileentity.TileEntityType
import net.minecraft.util.Direction
import net.minecraft.util.text.ITextComponent
import net.minecraft.util.text.TranslationTextComponent
import szewek.flux.FTiles
import szewek.flux.FTiles.TileType
import szewek.flux.recipe.AbstractMachineRecipe
import szewek.flux.recipe.AlloyingRecipe
import java.util.function.Function

class Machine2For1Tile(
        typeIn: TileEntityType<*>?,
        recipeTypeIn: IRecipeType<out AbstractMachineRecipe?>?,
        factory: MenuFactory,
        private val titleId: String
) : AbstractMachineTile(typeIn!!, recipeTypeIn!!, factory, 2, 1) {
    override fun getSlotsForFace(side: Direction) = when (side) {
        Direction.UP -> SLOTS_UP
        Direction.DOWN -> SLOTS_DOWN
        else -> SLOTS_SIDE
    }

    override fun getDefaultName() = TranslationTextComponent(titleId)

    companion object {
        private val SLOTS_UP = intArrayOf(0)
        private val SLOTS_SIDE = intArrayOf(1)
        private val SLOTS_DOWN = intArrayOf(2)

        @JvmStatic
        fun make(recipeType: IRecipeType<out AbstractMachineRecipe?>?, factory: MenuFactory, titleName: String): (TileType<Machine2For1Tile>) -> Machine2For1Tile {
            val titleId = "container.flux.$titleName"
            return { type -> Machine2For1Tile(type, recipeType, factory, titleId) }
        }
    }

}